aws/rust-runtime/aws-config/src/imds/credentials.rs

```diff
@@ -207,28 +319,58 @@ impl ImdsCredentialsProvider {
     }
 
     async fn retrieve_credentials(&self) -> provider::Result {
-        if self.imds_disabled() {
-            let err = format!(
-                "IMDS disabled by {} env var set to `true`",
-                super::env::EC2_METADATA_DISABLED
-            );
+        if let (true, origin) = self.imds_disabled().await {
+            let err = format!("IMDS disabled by {origin} set to `true`",);
             tracing::debug!(err);
             return Err(CredentialsError::not_loaded(err));
         }
+
         tracing::debug!("loading credentials from IMDS");
-        let profile: Cow<'_, str> = match &self.profile {
-            Some(profile) => profile.into(),
-            None => self.get_profile_uncached().await?.into(),
-        };
+
+        let profile = self.resolve_profile_name().await?;
         tracing::debug!(profile = %profile, "loaded profile");
-        let credentials = self
+
+        let credentials = match self
             .client
-            .get(format!(
-                "/latest/meta-data/iam/security-credentials/{}",
-                profile
-            ))
+            .get(format!("{uri_base}{profile}", uri_base = self.uri_base()))
             .await
-            .map_err(CredentialsError::provider_error)?;
+        {
+            Ok(credentials) => {
+                let state = &mut self.provider_state.write().unwrap();
+                state.resolved_profile = Some(profile.to_string());
```

Comment:
isn't this set in `resolve_profile_name()` already?

Reply:
Ah, thanks for catching. Came from stale pseudo-code in the SEP (pseudo-code has already been updated correctly).